prow/resultstore/payload.go

```diff
@@ -0,0 +1,242 @@
+	case v1.AbortedState:
+		status = resultstore.Status_CANCELLED
+	case v1.ErrorState:
+		status = resultstore.Status_INCOMPLETE
+	}
+	return &resultstore.StatusAttributes{
+		Status: status,
+	}
+}
+
+func (p *Payload) invocationTiming() *resultstore.Timing {
+	start := p.Job.Status.StartTime.Time
+	duration := p.Job.Status.CompletionTime.Time.Sub(start)
+	return &resultstore.Timing{
+		StartTime: &timestamppb.Timestamp{
+			Seconds: start.Unix(),
+		},
+		Duration: &durationpb.Duration{
+			Seconds: int64(duration.Seconds()),
+		},
+	}
+}
+
+func (p *Payload) invocationAttributes() *resultstore.InvocationAttributes {
+	return &resultstore.InvocationAttributes{
+		// TODO: ProjectID might be assigned directly from the GCS
+		// BucketAttrs.ProjectNumber; requires a raw GCS client.
+		ProjectId:   p.ProjectID,
+		Labels:      []string{"prow"},
+		Description: descriptionFromLabels(p.Job.Labels),
+	}
+}
+
+func descriptionFromLabels(labels map[string]string) string {
+	jt := labels[kube.ProwJobTypeLabel]
+	repo := labels[kube.RepoLabel]
+	pull := labels[kube.PullLabel]
+	if ps := labels[kube.GerritPatchset]; ps != "" {
+		pull += "/" + ps
+	}
+	buildID := labels[kube.ProwBuildIDLabel]
+	job := labels[kube.ContextAnnotation]
+	return fmt.Sprintf("%s for %s/%s/%s/%s", jt, repo, pull, buildID, job)
+}
+
+func (p *Payload) workspaceInfo() *resultstore.WorkspaceInfo {
+	return &resultstore.WorkspaceInfo{
+		CommandLines: commandLines(p.Job),
+	}
+}
+
+func commandLines(pj *v1.ProwJob) []*resultstore.CommandLine {
+	var cl []*resultstore.CommandLine
+	if pj.Spec.PodSpec != nil {
+		for _, c := range pj.Spec.PodSpec.Containers {
+			cl = append(cl, &resultstore.CommandLine{
+				Label:   c.Name,
+				Args:    c.Args,
+				Command: strings.Join(c.Command, " "),
+			})
+		}
+	}
+	return cl
+}
+
+func (p *Payload) invocationProperties() []*resultstore.Property {
+	ps := []*resultstore.Property{
+		{
+			Key:   "Instance",
+			Value: p.Job.Status.BuildID,
+		},
+		{
+			Key:   "Job",
+			Value: p.Job.Spec.Job,
+		},
+		{
+			Key:   "Prow_Dashboard_URL",
+			Value: p.Job.Status.URL,
+		},
+		{
+			Key:   "Repo",
+			Value: p.Job.Spec.Refs.RepoLink,
+		},
+	}
+	if p.Started != nil {
+		ps = append(ps, &resultstore.Property{
+			Key:   "Commit",
+			Value: p.Started.RepoCommit,
+		})
+		for _, branch := range p.Started.Repos {
+			ps = append(ps, &resultstore.Property{
+				Key:   "Branch",
+				Value: branch,
+			})
+		}
+	}
+	if p.Job.Spec.PodSpec != nil {
+		for _, c := range p.Job.Spec.PodSpec.Containers {
+			for _, e := range c.Env {
+				ps = append(ps, &resultstore.Property{
+					Key:   "Env",
+					Value: e.Name + "=" + e.Value,
+				})
+			}
+		}
+	}
+	return ps
+}
+
+func (p *Payload) defaultConfiguration() *resultstore.Configuration {
+	return &resultstore.Configuration{
+		Id: &resultstore.Configuration_Id{
+			ConfigurationId: "default",
+		},
+	}
+}
+
+func (p *Payload) targetID() string {
+	if p.Job == nil {
+		return "Unknown"
+	}
```

Comment:
nit: Elsewhere we use `p.Job` without checking if it is nil. We should be consistent about this.

Reply:
I'm not finding where we do. Note that the helper functions above for building an Invocation are all guarded by the check in p.invocation().

Note the unit tests exercise these methods with a nil Job (and other nil parts).

In the spirit of nil pointers, I notice now that Spec.Refs and Status.CompletionTime (but not StartTime) are also pointers. Expanded test cases and updated to guard these.